### PR TITLE
MDEV-39223 Reversed executable comments incorrectly written in binlog

### PR DESCRIPTION
**Pull request by tonychen2001**
## Description

This addresses an issue introduced in #4724.

When a reversed executable comment (/*!!version */ or /*M!!version */)
is not executed on the master (server version >= specified version),
the binlog writer only neutralized the second '!' to a space, leaving
`/*! command */`, an executable comment the slave would then try to
execute.

Fix (non-executed path): also replace the first '!' with a space so the
SQL becomes /*  (a plain comment). The restore path for unclosed comments
also restores both '!' characters.

Additionally, when a reversed executable comment IS executed on the
source (server version < specified version), the /*!!version syntax was
preserved verbatim in the binlog. A replica with a higher version could
re-evaluate the reversed condition, fail it, and skip the command —
causing source/replica divergence. Unlike forward executable comments
(/*!version */), where replication to same-or-higher versions guarantees
the replica also executes, reversed comments have the opposite semantics
and are not safe to preserve as-is.

Fix (executed path): overwrite the version digits with spaces in the raw
query buffer so the binlog contains /*!! (a non-versioned reversed
executable comment) which the replica always executes regardless of its
version.

Ultimately, if a reverse comment was executed on the source it must be executed on the replica regardless of
versioning. Similarly, if a reverse comment was **not** executed on the source it must **not** be executed on the replica regardless of versioning.

Additionally, we mark queries that are patched in-place as non-cacheable when `query_cache_strip_comments=OFF`. 
The lexer rewrites the raw query buffer when it takes a reversed executable
comment (digits -> spaces) or skips a forward/reversed comment (the '!' ->
space). With `query_cache_strip_comments=OFF` the cache key reads from this
buffer, so the lookup key (built before parsing) and the store key (built
after) disagree: the entry is inserted but never hit. Mark such queries as not
safe to cache to avoid polluting the cache. When `query_cache_strip_comments=ON`
caching remains enabled since its key comes from a separate comment-stripped
copy of the query buffer.

## How can this PR be tested?

Added rpl_reversed_comments.test mirroring rpl_conditional_comments.test
with equivalent test cases:
- mix of applied/not-applied
- prepared statements
- unclosed comments,
- nested comments
- delimiter edge case
- MariaDB-specific /*M!! syntax

Also added `main.query_cache_executable_comments.result` for testing query cache behavior.

## Basing the PR against the correct MariaDB version

* [x] _This is a bug fix, and the PR is based against the branch main as the 13.0 branch does not exist at the moment._

## Copyright

All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

### CODE REVIEW

**Review by gkodinov — COMMENTED** [review]
Thank you for fixing this so fast! This is a preliminary review. 

LGTM. 

Please stand by for the final review.

**Review by gkodinov — APPROVED** [review]

File: mysql-test/suite/rpl/t/rpl_reversed_comments.test [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#  - Use ' ' instead of '!' in the conditional comments which are not applied on
+#  master. So they become common comments and will not be applied on slave.
+#
+#  - Example: 
+#  'INSERT INTO t1 VALUES (1) /*!!10000, (2)*/ /*!!999999 ,(3)*/ 
+#  will be binlogged as 
+#  'INSERT INTO t1 VALUES (1) /*  10000, (2)*/ /*!!999999 ,(3)*/'.
+###############################################################################
+source include/have_binlog_format_statement.inc;
+source include/master-slave.inc;
```
Comment (bnestere) [inline mysql-test/suite/rpl/t/rpl_reversed_comments.test:16]: The test itself is well constructed and tests a variety of use cases, which is great. Though for logic which changes binlogging, it is best practice to have the test use chain replication to ensure nothing gets lost in translation from a middle slave. So please make this

```
--let $rpl_topology=1->2, 2->3
--source include/rpl_init.inc
```

and use `connection server_1` and `connection server_2` instead of `connection master` and `connection slave`.

You probably also have to manually tag points to sync at and who to sync with.

So instead of 
```
--sync_slave_with_master
```

you'd do
```
--connection server_1
--source include/save_master_gtid.inc
--connection server_3
--source include/sync_with_master_gtid.inc
```

Note you can skip server 2 in-between to help with conciseness.

File: sql/sql_lex.cc [hunk 2]
```diff
@@ -2479,6 +2479,17 @@ int Lex_input_stream::lex_one_token(YYSTYPE *yylval, THD *thd)
...
+            if (reversed_comment)
+            {
+              /*
+                Overwrite the version digits with spaces in the raw query
+                buffer so the binlog contains a non-versioned reversed
+                executable comment that the replica always executes.
+              */
+              char *p= (char *) get_ptr();
+              for (uint i= 0; i < length; i++)
+                p[i]= ' ';
```
Comment (bnestere) [inline sql/sql_lex.cc:2491]: This looks to go with your comment for

> Additionally, when a reversed executable comment IS executed on the
source (server version < specified version), the /!!version syntax was
preserved verbatim in the binlog. A replica with a higher version could
re-evaluate the reversed condition, fail it, and skip the command —
causing source/replica divergence. Unlike forward executable comments
(/!version */), where replication to same-or-higher versions guarantees
the replica also executes, reversed comments have the opposite semantics
and are not safe to preserve as-is.
>
> ...
>
> Ultimately, if a reverse comment was executed on the source it must be executed on the replica regardless of
versioning. Similarly, if a reverse comment was not executed on the source it must not be executed on the replica regardless of versioning.

Though I'm not convinced this is the correct behavior. In a far-away future where 13.X is long EoL and features are very different; I could see a user _wanting_ to write a SQL command which needs to execute on some old (but still has reversed executable comments) master, but the upgraded slave would have some new syntax the user would prefer to use. 

@vuvova do you have any thoughts?

**Review by bnestere — CHANGES_REQUESTED** [review]
Thank you for the patch, @tonychen2001 !

i've left a couple comments.

File: sql/sql_lex.cc (hunk 2 above)
Comment (vuvova) [reply sql/sql_lex.cc:2491]: Yes, this was the behavior I suggested.

To arrive at the desired behavior we need a use case. I came up with something based on what Federico wrote in [MDEV-7381](https://jira.mariadb.org/browse/MDEV-7381) (version numbers below are not necessarily correct, they're just examples):
```sql
/*!!10200 DROP TABLE IF EXISTS t1; */
CREATE /*!10200 OR REPLACE*/ TABLE /*!!10200 IF NOT EXISTS */ t1 (a int)
```
here it is obvious that if a conditional comment is executed, the reversed conditional comment must not be. And vice versa. In no case it is correct to execute both or none.

and because the slave behavior towards conditional comments is the same as master's — a comment is either executed on both or on neither, the same must be true for reversed comments as well, otherwise you'll arrive at the discrepancy when a slave could execute both comments or none.

Comment (bnestere) [reply sql/sql_lex.cc:2491]: Ok I see.

Then @tonychen2001 I think your goal

> Fix (executed path): overwrite the version digits with spaces in the raw
query buffer so the binlog contains /*!! (a non-versioned reversed
executable comment) which the replica always executes regardless of its
version.

is not honored in your patch (or perhaps I misunderstand how it is to be used).

The command:

```
INSERT INTO t1 VALUES(/*M!140000 110 + */ 1), (/*M!!140000 220 + */ 2);
```

results in
```
select * from t1;
c1
1
122
```

and is binlogged as
```
INSERT INTO t1 VALUES(/*M 140000 110 + */ 1), (/*M!!140000 120 + */ 2)
```

The version isn't overridden in the binlogged version, so when some 14.X slave comes around, it will have data divergence.
```
select * from t1;
c1
1
2
```

Comment (tonychen2001) [reply sql/sql_lex.cc:2491]: Hi @bnestere , I wasn't able to reproduce the problem. Could you let me know if I am doing something wrong

```
docker exec mariadb bash -c 'cd /quick-rebuilds && ./build/client/mariadb -u root --socket=/tmp/mysql.sock -t test << "EOF"
  RESET MASTER;
  CREATE TABLE t1(c1 INT);
  INSERT INTO t1 VALUES(/*M!140000 110 + */ 1), (/*M!!140000 220 + */ 2);
  SELECT * FROM t1;
  SHOW BINLOG EVENTS;
  DROP TABLE t1;
  EOF'

+------+
| c1   |
+------+
|    1 |
|  222 |
+------+
+---------------+-----+-------------------+-----------+-------------+------------------------------------------------------------------------------------+
| Log_name      | Pos | Event_type        | Server_id | End_log_pos | Info                                                                               |
+---------------+-----+-------------------+-----------+-------------+------------------------------------------------------------------------------------+
| binlog.000001 |   4 | Format_desc       |         1 |         257 | Server ver: 13.0.1-MariaDB-debug-log, Binlog ver: 4                                |
| binlog.000001 | 257 | Gtid_list         |         1 |         286 | []                                                                                 |
| binlog.000001 | 286 | Binlog_checkpoint |         1 |         326 | binlog.000001                                                                      |
| binlog.000001 | 326 | Gtid              |         1 |         368 | GTID 0-1-1                                                                         |
| binlog.000001 | 368 | Query             |         1 |         467 | use `test`; CREATE TABLE t1(c1 INT)                                                |
| binlog.000001 | 467 | Gtid              |         1 |         509 | BEGIN GTID 0-1-2                                                                   |
| binlog.000001 | 509 | Query             |         1 |           0 | use `test`; INSERT INTO t1 VALUES(/*M 140000 110 + */ 1), (/*M!!       220 + */ 2) |
| binlog.000001 | 646 | Xid               |         1 |         677 | COMMIT /* xid=60 */                                                                |
+---------------+-----+-------------------+-----------+-------------+------------------------------------------------------------------------------------+
```

Comment (bnestere) [reply sql/sql_lex.cc:2491]: Strange, I don't see that behavior anymore either. Perhaps I was accidentally testing in the wrong directory, _or something_.

**Review by bnestere — APPROVED** [review]
Looks good, @tonychen2001! Thanks for updating the test with a quick turnaround. 

Please also remember to update the git commit message with your reviewer (me)

```
Reviewed-by: Brandon Nesterenko <brandon.nesterenko@mariadb.com>
```

File: mysql-test/suite/rpl/t/rpl_reversed_comments.test (hunk 1 above)
Comment (tonychen2001) [reply mysql-test/suite/rpl/t/rpl_reversed_comments.test]: Updated the MTR to test chained replication

**Comment by vuvova** [comment]
@tonychen2001, would you also like to fix [MDEV-39421](https://jira.mariadb.org/browse/MDEV-39421) ? After all it was discovered in your branch. Just mark any query where a conditional comment was patched in-place as non-cacheable.

**Comment by tonychen2001** [comment]
> @tonychen2001, would you also like to fix [MDEV-39421](https://jira.mariadb.org/browse/MDEV-39421) ? After all it was discovered in your branch. Just mark any query where a conditional comment was patched in-place as non-cacheable.

Yes of course. 

Btw, it may be a couple weeks before I'm able to get to it as I'm going to be taking a couple weeks off starting April 28 and I'm caught up with some other work I'm trying to wrap up before then.

**Comment by tonychen2001** [comment]
Hi @vuvova, I have added a commit to address [MDEV-39421](https://jira.mariadb.org/browse/MDEV-39421). I wasn't sure how to go about creating a new PR since the changes are quite related to each other so I just added it to this PR.

Please let me know your thoughts.

**Comment by gkodinov** [comment]
FWIW, this PR aims to resolve both https://jira.mariadb.org/browse/MDEV-39223 and https://jira.mariadb.org/browse/MDEV-39421

**Review by vuvova — APPROVED** [review]
